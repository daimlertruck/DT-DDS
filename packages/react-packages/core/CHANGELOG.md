# @dt-ui/react-core

## 0.1.0-beta.13

### Patch Changes

- fix: match accordion design

## 0.1.0-beta.12

### Patch Changes

- fix: make icons an optional property in theme

## 0.1.0-beta.11

### Patch Changes

- refactor: extract text-field to new package with greenlane theme

## 0.1.0-beta.10

### Patch Changes

- refactor: move and refactor checkbox refactor checkbox according to design

## 0.1.0-beta.9

### Patch Changes

- fix(icons): export css and update imports on deps

## 0.1.0-beta.8

### Patch Changes

- chore: move util functions to the core

## 0.1.0-beta.7

### Minor Changes

- feat: add contextual color add content tertiary to palette

## 0.1.0-beta.6

### Minor Changes

- feat: create Icons package
- feat: add icons to theme Provider

### Patch Changes

- Updated dependencies
  - @dt-ui/icons@0.1.0-beta.1

## 0.1.0-beta.5

### Minor Changes

- feat: create TruckAPI theme and update default

## 0.1.0-beta.4

### Minor Changes

- feat: update contextual color to follow figma spec

## 0.1.0-beta.3

### Minor Changes

- feat: review Message designs based on designs

## 0.1.0-beta.2

### Patch Changes

- chore: migrate jest setup file to jest package
- chore: migrate jest configuration to typescript

## 0.1.0-beta.1

### Patch Changes

- chore: create core package
