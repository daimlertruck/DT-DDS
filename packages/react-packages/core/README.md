# Core Package

The Core Package provides a centralized collection of shared utilities, hooks, components, constants and types. This package is designed to enhance the development experience by promoting code reusability and maintaining consistency across various components.

## Stack

- [TypeScript](https://www.typescriptlang.org/) for static type checking
- [React](https://reactjs.org/) — JavaScript library for user interfaces
- [Emotion](https://emotion.sh/docs/introduction) — for writing css styles with JavaScript
- [Storybook](https://storybook.js.org/) — UI component environment powered by Vite
- [Jest](https://jestjs.io/) - JavaScript Testing Framework
- [React Testing Library](https://testing-library.com/) - to test UI components in a user-centric way
- [ESLint](https://eslint.org/) for code linting
- [Prettier](https://prettier.io) for code formatting
- [Tsup](https://github.com/egoist/tsup) — TypeScript bundler powered by esbuild
- [Yarn](https://yarnpkg.com/) from managing packages

## Commands

- `yarn build` - Build the package
- `yarn dev` - Run the package locally
- `yarn lint` - Lint all files within this package
- `yarn test` Run all unit tests

## Compilation

Running `yarn build` from the root of the package will use [tsup](https://tsup.egoist.dev/) to compile the raw TypeScript and React code to plain JavaScript.

The `/dist` folder contains the compiled output.

```bash
core
└── dist
    ├── index.d.ts  <-- Types
    ├── index.js    <-- CommonJS version
    └── index.mjs   <-- ES Modules version
    ...
```

## Versioning

Follows [semantic versioning](https://semver.org/)

## &copy; License

Licensed under [MIT License](LICENSE.md)
