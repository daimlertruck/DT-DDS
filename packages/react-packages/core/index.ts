import { CustomTheme } from './src/types';

export * from './src';

declare module '@emotion/react' {
  export interface Theme extends CustomTheme {}
}
