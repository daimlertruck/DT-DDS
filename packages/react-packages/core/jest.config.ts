import type { Config } from 'jest';
import baseConfig from 'jest-config';

const config: Config = {
  ...baseConfig,
};

export default config;
