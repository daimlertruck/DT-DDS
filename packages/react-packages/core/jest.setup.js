import '@testing-library/jest-dom';
import { configure } from '@testing-library/react';
import './jest-mock-globals';

configure({ defaultHidden: true });
