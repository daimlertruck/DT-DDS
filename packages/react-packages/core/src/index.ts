export * from './components';
export * from './constants';
export * from './hooks';
export * from './themes';
export type {
  BaseProps,
  Colors,
  ComponentSize,
  CustomTheme as Theme,
  FontFace,
  Orientation,
  PaletteKeys,
} from './types';
export {
  debounce,
  hexToRgba,
  useClickOutside,
  withProviders,
  renderStatic,
  getContextualColor,
  getTextStyles,
} from './utils';
