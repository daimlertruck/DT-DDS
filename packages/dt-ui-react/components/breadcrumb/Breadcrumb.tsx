import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { ArrowRightIcon } from '../../core';
import { BaseProps } from '../../types';
import { Tooltip } from '../tooltip';

import {
  BreadcrumbStyled,
  ListStyled,
  ItemStyled,
  ItemContainerStyled,
} from './Breadcrumb.styled';
import { useCollapsedBreadcrumb } from './hooks/useCollapsedBreadcrumb';

const BREADCRUMB_CHARACTER_LIMIT = 30;

type BreadcrumbSeparator = 'slash' | 'arrow';

export interface BreadcrumbProps extends BaseProps {
  separator?: BreadcrumbSeparator;
}

export const Breadcrumb = ({
  children,
  separator = 'slash',
}: BreadcrumbProps) => {
  const { visibleChildren, breadcrumbRef, breadcrumbListRef } =
    useCollapsedBreadcrumb(children);

  return (
    <BreadcrumbStyled
      ref={breadcrumbRef}
      aria-label='Breadcrumb'
      data-testid='breadcrumb'
    >
      <ListStyled ref={breadcrumbListRef}>
        {withSeparator(visibleChildren, separator)}
      </ListStyled>
    </BreadcrumbStyled>
  );
};

const BreadcrumbItem = ({ children, style }: BaseProps) => {
  const { truncatedChildren, tooltip } = mapTruncateChildren(children);

  return (
    <ItemStyled style={style}>
      <Tooltip style={{ overflow: 'hidden' }}>
        <ItemContainerStyled>{truncatedChildren}</ItemContainerStyled>
        <Tooltip.Content>{tooltip}</Tooltip.Content>
      </Tooltip>
    </ItemStyled>
  );
};

const truncateText = (text: string) => {
  if (text.length >= BREADCRUMB_CHARACTER_LIMIT) {
    return `${text.substring(0, BREADCRUMB_CHARACTER_LIMIT)}...`;
  }
  return text;
};

const mapTruncateChildren = (
  children: ReactNode
): { tooltip: string; truncatedChildren: ReactNode } => {
  let tooltip = '';
  return {
    truncatedChildren: Children.map(children, (child) => {
      if (typeof child === 'string') {
        const truncatedText = truncateText(child);
        tooltip = truncatedText !== child ? child : '';

        return truncatedText;
      } else if (isValidElement(child)) {
        const { truncatedChildren, tooltip: childTooltip } =
          mapTruncateChildren(child.props.children);
        tooltip = childTooltip;

        return cloneElement(child, {
          ...child.props,
          children: truncatedChildren,
        });
      } else {
        return null;
      }
    }),
    tooltip,
  };
};

const withSeparator = (children: ReactNode, separator: BreadcrumbSeparator) => {
  const addSeparator = () =>
    ({
      slash: <span>/</span>,
      arrow: <ArrowRightIcon />,
    }[separator]);

  const isLastBreadcrumb = (index: number) =>
    Children.count(children) - 1 === index;

  return Children.map(children, (child, index) => (
    <>
      {child}
      {!isLastBreadcrumb(index) && addSeparator()}
    </>
  ));
};

Breadcrumb.Item = BreadcrumbItem;
