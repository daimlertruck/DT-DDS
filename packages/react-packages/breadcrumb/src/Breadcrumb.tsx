import { BaseProps } from '@dt-dds/react-core';
import { Tooltip } from '@dt-dds/react-tooltip';
import { Children, ReactNode, cloneElement, isValidElement } from 'react';

import { ArrowRightIcon } from '../../../dt-dds-react/core';

import {
  BreadcrumbStyled,
  ListStyled,
  ItemStyled,
  ItemContainerStyled,
} from './Breadcrumb.styled';
import { useCollapsedBreadcrumb } from './hooks/useCollapsedBreadcrumb';

const BREADCRUMB_CHARACTER_LIMIT = 30;

type BreadcrumbSeparator = 'slash' | 'arrow';

export interface BreadcrumbProps extends BaseProps {
  separator?: BreadcrumbSeparator;
}

export const Breadcrumb = ({
  children,
  separator = 'slash',
}: BreadcrumbProps) => {
  const { visibleChildren, breadcrumbRef, breadcrumbListRef } =
    useCollapsedBreadcrumb(children);

  return (
    <BreadcrumbStyled
      aria-label='Breadcrumb'
      data-testid='breadcrumb'
      ref={breadcrumbRef}
    >
      <ListStyled ref={breadcrumbListRef}>
        {withSeparator(visibleChildren, separator)}
      </ListStyled>
    </BreadcrumbStyled>
  );
};

const BreadcrumbItem = ({ children, style }: BaseProps) => {
  const { truncatedChildren, tooltip } = mapTruncateChildren(children);

  return (
    <ItemStyled style={style}>
      <Tooltip style={{ overflow: 'hidden' }}>
        <ItemContainerStyled>{truncatedChildren}</ItemContainerStyled>
        <Tooltip.Content>{tooltip}</Tooltip.Content>
      </Tooltip>
    </ItemStyled>
  );
};

const truncateText = (text: string) => {
  if (text.length >= BREADCRUMB_CHARACTER_LIMIT) {
    return `${text.substring(0, BREADCRUMB_CHARACTER_LIMIT)}...`;
  }
  return text;
};

const mapTruncateChildren = (
  children: ReactNode
): { tooltip: string; truncatedChildren: ReactNode } => {
  let tooltip = '';
  return {
    truncatedChildren: Children.map(children, (child) => {
      if (typeof child === 'string') {
        const truncatedText = truncateText(child);
        tooltip = truncatedText !== child ? child : '';

        return truncatedText;
      } else if (isValidElement(child)) {
        const { truncatedChildren, tooltip: childTooltip } =
          mapTruncateChildren(child.props.children);
        tooltip = childTooltip;

        return cloneElement(child, {
          ...child.props,
          children: truncatedChildren,
        });
      } else {
        return null;
      }
    }),
    tooltip,
  };
};

const withSeparator = (children: ReactNode, separator: BreadcrumbSeparator) => {
  const addSeparator = () =>
    ({
      slash: <span>/</span>,
      arrow: <ArrowRightIcon />,
    }[separator]);

  const isLastBreadcrumb = (index: number) =>
    Children.count(children) - 1 === index;

  return Children.map(children, (child, index) => (
    <>
      {child}
      {!isLastBreadcrumb(index) && addSeparator()}
    </>
  ));
};

Breadcrumb.Item = BreadcrumbItem;
