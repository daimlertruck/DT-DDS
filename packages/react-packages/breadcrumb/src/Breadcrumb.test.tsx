import { withProviders } from '@dt-dds/react-core';
import { render } from '@testing-library/react';

import { ViewAgendaIcon } from '../../../dt-dds-react/core';

import { Breadcrumb } from './Breadcrumb';

describe('<Breadcrumb /> component', () => {
  const ProvidedBreadcrumb = withProviders(Breadcrumb);

  beforeEach(() => {
    jest.mock('@dt-dds/react-core', () => ({
      useMedia: jest.fn(() => true),
    }));
  });

  it('renders nav element with the correct breadcrumb items', () => {
    const breadcrumbs = [
      { name: 'Name 1', showIcon: true },
      {
        name: 'Second page has a long text to describe its path',
      },
      { disabled: true, name: 'Name 3' },
      {
        name: 'Last page has a long text to describe its path',
      },
    ];

    const { container } = render(
      <ProvidedBreadcrumb separator='arrow'>
        {breadcrumbs.map((breadcrumb) => (
          <Breadcrumb.Item key={breadcrumb.name}>
            {breadcrumb.showIcon ? <ViewAgendaIcon /> : null}
            <a href='/'>{breadcrumb.name}</a>
          </Breadcrumb.Item>
        ))}
      </ProvidedBreadcrumb>
    );

    expect(container).toMatchSnapshot();
  });
});
