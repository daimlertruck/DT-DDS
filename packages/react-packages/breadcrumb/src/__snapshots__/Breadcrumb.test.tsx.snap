// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumb /> component renders nav element with the correct breadcrumb items 1`] = `
.emotion-0 {
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #636464;
  margin-bottom: 12px;
  overflow: hidden;
}

.emotion-2 {
  padding: 0px;
  margin: 0px;
  width: 100%;
  list-style: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  white-space: nowrap;
  gap: 8px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-4:not(:last-child) {
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  color: #008DFC;
  cursor: pointer;
}

.emotion-4:not(:last-child) svg {
  color: #008DFC;
}

.emotion-4:not(:last-child) * {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.emotion-4:not(:last-child):hover {
  color: #025497;
}

.emotion-4:not(:last-child):hover svg {
  color: #025497;
}

.emotion-4 a {
  color: inherit;
}

.emotion-4 svg {
  color: #636464;
}

.emotion-4:last-child * {
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.emotion-6 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
}

<div>
  <nav
    aria-label="Breadcrumb"
    class="emotion-0 emotion-1"
    data-testid="breadcrumb"
  >
    <ul
      class="emotion-2 emotion-3"
    >
      <li
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          data-testid="tooltip-container"
          style="overflow: hidden;"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <svg
              fill="none"
              height="17"
              viewBox="0 0 16 17"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12.6667 9.33329H3.33333C2.6 9.33329 2 9.93329 2 10.6666V13.3333C2 14.0666 2.6 14.6666 3.33333 14.6666H12.6667C13.4 14.6666 14 14.0666 14 13.3333V10.6666C14 9.93329 13.4 9.33329 12.6667 9.33329ZM12.6667 13.3333H3.33333V10.6666H12.6667V13.3333Z"
                fill="currentColor"
              />
              <path
                d="M12.6667 2.66663H3.33333C2.6 2.66663 2 3.26663 2 3.99996V6.66663C2 7.39996 2.6 7.99996 3.33333 7.99996H12.6667C13.4 7.99996 14 7.39996 14 6.66663V3.99996C14 3.26663 13.4 2.66663 12.6667 2.66663ZM12.6667 6.66663H3.33333V3.99996H12.6667V6.66663Z"
                fill="currentColor"
              />
            </svg>
            <a
              href="/"
            >
              Name 1
            </a>
          </div>
        </div>
      </li>
      <svg
        fill="none"
        height="12"
        viewBox="0 0 8 12"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.99468 4.99507C7.18215 5.1826 7.28747 5.4369 7.28747 5.70207C7.28747 5.96723 7.18215 6.22154 6.99468 6.40907L2.40168 11.0021C2.21539 11.1846 1.96458 11.2863 1.70375 11.285C1.44292 11.2837 1.19315 11.1795 1.00871 10.995C0.824278 10.8106 0.72008 10.5608 0.718763 10.3C0.717445 10.0392 0.819115 9.78836 1.00168 9.60207L4.90168 5.70207L1.00168 1.80207C0.819115 1.61578 0.717445 1.36497 0.718763 1.10414C0.72008 0.843312 0.824278 0.593541 1.00871 0.409105C1.19315 0.224668 1.44292 0.120471 1.70375 0.119153C1.96458 0.117836 2.21539 0.219506 2.40168 0.402069L6.99468 4.99507Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <li
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          data-testid="tooltip-container"
          style="overflow: hidden;"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <a
              href="/"
            >
              Second page has a long text to...
            </a>
          </div>
        </div>
      </li>
      <svg
        fill="none"
        height="12"
        viewBox="0 0 8 12"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.99468 4.99507C7.18215 5.1826 7.28747 5.4369 7.28747 5.70207C7.28747 5.96723 7.18215 6.22154 6.99468 6.40907L2.40168 11.0021C2.21539 11.1846 1.96458 11.2863 1.70375 11.285C1.44292 11.2837 1.19315 11.1795 1.00871 10.995C0.824278 10.8106 0.72008 10.5608 0.718763 10.3C0.717445 10.0392 0.819115 9.78836 1.00168 9.60207L4.90168 5.70207L1.00168 1.80207C0.819115 1.61578 0.717445 1.36497 0.718763 1.10414C0.72008 0.843312 0.824278 0.593541 1.00871 0.409105C1.19315 0.224668 1.44292 0.120471 1.70375 0.119153C1.96458 0.117836 2.21539 0.219506 2.40168 0.402069L6.99468 4.99507Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <li
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          data-testid="tooltip-container"
          style="overflow: hidden;"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <a
              href="/"
            >
              Name 3
            </a>
          </div>
        </div>
      </li>
      <svg
        fill="none"
        height="12"
        viewBox="0 0 8 12"
        width="8"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M6.99468 4.99507C7.18215 5.1826 7.28747 5.4369 7.28747 5.70207C7.28747 5.96723 7.18215 6.22154 6.99468 6.40907L2.40168 11.0021C2.21539 11.1846 1.96458 11.2863 1.70375 11.285C1.44292 11.2837 1.19315 11.1795 1.00871 10.995C0.824278 10.8106 0.72008 10.5608 0.718763 10.3C0.717445 10.0392 0.819115 9.78836 1.00168 9.60207L4.90168 5.70207L1.00168 1.80207C0.819115 1.61578 0.717445 1.36497 0.718763 1.10414C0.72008 0.843312 0.824278 0.593541 1.00871 0.409105C1.19315 0.224668 1.44292 0.120471 1.70375 0.119153C1.96458 0.117836 2.21539 0.219506 2.40168 0.402069L6.99468 4.99507Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
      <li
        class="emotion-4 emotion-5"
      >
        <div
          class="emotion-6 emotion-7"
          data-testid="tooltip-container"
          style="overflow: hidden;"
        >
          <div
            class="emotion-8 emotion-9"
          >
            <a
              href="/"
            >
              Last page has a long text to d...
            </a>
          </div>
        </div>
      </li>
    </ul>
  </nav>
</div>
`;
